// mongodb-memory-server-core/lib/MongoMemoryReplSet.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var MongoMemoryServer_1 = __importDefault(require("./MongoMemoryServer"));
var db_util_1 = require("./util/db_util");
var debug_1 = __importDefault(require("debug"));
var mongodb_1 = require("mongodb");
var log = debug_1.default('MongoMS:MongoMemoryReplSet');
var MongoMemoryReplSet = /** @class */ (function (_super) {
    __extends(MongoMemoryReplSet, _super);
    function MongoMemoryReplSet(opts) {
        if (opts === void 0) { opts = {}; }
        var _this = _super.call(this) || this;
        _this.servers = [];
        var replSetDefaults = {
            auth: false,
            args: [],
            name: 'testset',
            count: 1,
            dbName: db_util_1.generateDbName(),
            ip: '127.0.0.1',
            oplogSize: 1,
            spawn: {},
            storageEngine: 'ephemeralForTest',
            configSettings: {},
        };
        _this._state = 'stopped';
        _this.opts = {
            binary: opts.binary || {},
            instanceOpts: opts.instanceOpts || [],
            replSet: __assign(__assign({}, replSetDefaults), opts.replSet),
        };
        if (!_this.opts.replSet.args)
            _this.opts.replSet.args = [];
        _this.opts.replSet.args.push('--oplogSize', "" + _this.opts.replSet.oplogSize);
        if (!(opts && opts.autoStart === false)) {
            log('Autostarting MongoMemoryReplSet.');
            setTimeout(function () { return _this.start(); }, 0);
        }
        process.on('beforeExit', function () { return _this.stop(); });
        return _this;
    }
    MongoMemoryReplSet.prototype.getConnectionString = function (otherDb) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.getUri(otherDb)];
            });
        });
    };
    /**
     * Returns database name.
     */
    MongoMemoryReplSet.prototype.getDbName = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // this function is only async for consistency with MongoMemoryServer
                // I don't see much point to either of them being async but don't
                // care enough to change it and introduce a breaking change.
                return [2 /*return*/, this.opts.replSet.dbName];
            });
        });
    };
    /**
     * Returns instance options suitable for a MongoMemoryServer.
     * @param {MongoMemoryInstancePropBaseT} baseOpts
     */
    MongoMemoryReplSet.prototype.getInstanceOpts = function (baseOpts) {
        if (baseOpts === void 0) { baseOpts = {}; }
        var rsOpts = this.opts.replSet;
        var opts = {
            auth: !!rsOpts.auth,
            args: rsOpts.args,
            dbName: rsOpts.dbName,
            ip: rsOpts.ip,
            replSet: rsOpts.name,
            storageEngine: rsOpts.storageEngine,
        };
        if (baseOpts.args)
            opts.args = (rsOpts.args || []).concat(baseOpts.args);
        if (baseOpts.port)
            opts.port = baseOpts.port;
        if (baseOpts.dbPath)
            opts.dbPath = baseOpts.dbPath;
        if (baseOpts.storageEngine)
            opts.storageEngine = baseOpts.storageEngine;
        log('   instance opts:', opts);
        return opts;
    };
    /**
     * Returns a mongodb: URI to connect to a given database.
     */
    MongoMemoryReplSet.prototype.getUri = function (otherDb) {
        return __awaiter(this, void 0, void 0, function () {
            var dbName, ports, hosts;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this._state === 'init')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this._waitForPrimary()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        if (this._state !== 'running') {
                            throw new Error('Replica Set is not running. Use debug for more info.');
                        }
                        if (otherDb) {
                            dbName = typeof otherDb === 'string' ? otherDb : db_util_1.generateDbName();
                        }
                        else {
                            dbName = this.opts.replSet.dbName;
                        }
                        return [4 /*yield*/, Promise.all(this.servers.map(function (s) { return s.getPort(); }))];
                    case 3:
                        ports = _a.sent();
                        hosts = ports.map(function (port) { return "127.0.0.1:" + port; }).join(',');
                        return [2 /*return*/, "mongodb://" + hosts + "/" + dbName + "?replicaSet=" + this.opts.replSet.name];
                }
            });
        });
    };
    /**
     * Start underlying `mongod` instances.
     */
    MongoMemoryReplSet.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var servers, cnt, current, server;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        log('start');
                        if (this._state !== 'stopped') {
                            throw new Error("Already in 'init' or 'running' state. Use debug for more info.");
                        }
                        this.emit((this._state = 'init'));
                        log('init');
                        servers = this.opts.instanceOpts.map(function (opts) {
                            log('  starting server from instanceOpts:', opts, '...');
                            return _this._initServer(_this.getInstanceOpts(opts));
                        });
                        cnt = this.opts.replSet.count || 1;
                        current = 0;
                        while (servers.length < cnt) {
                            log("  starting server " + current + " of " + cnt + "...");
                            server = this._initServer(this.getInstanceOpts({}));
                            servers.push(server);
                        }
                        // ensures all servers are listening for connection
                        return [4 /*yield*/, Promise.all(servers.map(function (s) { return s.start(); }))];
                    case 1:
                        // ensures all servers are listening for connection
                        _a.sent();
                        this.servers = servers;
                        return [4 /*yield*/, this._initReplSet()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Stop the underlying `mongod` instance(s).
     */
    MongoMemoryReplSet.prototype.stop = function () {
        return __awaiter(this, void 0, void 0, function () {
            var servers;
            var _this = this;
            return __generator(this, function (_a) {
                if (this._state === 'stopped')
                    return [2 /*return*/, false];
                servers = this.servers;
                this.servers = [];
                return [2 /*return*/, Promise.all(servers.map(function (s) { return s.stop(); }))
                        .then(function () {
                        _this.emit((_this._state = 'stopped'));
                        return true;
                    })
                        .catch(function (err) {
                        log(err);
                        _this.emit((_this._state = 'stopped'), err);
                        return false;
                    })];
            });
        });
    };
    MongoMemoryReplSet.prototype.waitUntilRunning = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this._state === 'running')
                            return [2 /*return*/];
                        return [4 /*yield*/, new Promise(function (resolve) { return _this.once('running', function () { return resolve(); }); })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Connects to the first server from the list of servers and issues the `replSetInitiate`
     * command passing in a new replica set configuration object.
     */
    MongoMemoryReplSet.prototype._initReplSet = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uris, MongoClient, conn, db, admin, members, rsConfig, e_1, oldConfig;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this._state !== 'init') {
                            throw new Error('Not in init phase.');
                        }
                        log('Initializing replica set.');
                        if (!this.servers.length) {
                            throw new Error('One or more servers are required.');
                        }
                        return [4 /*yield*/, Promise.all(this.servers.map(function (server) { return server.getUri(); }))];
                    case 1:
                        uris = _a.sent();
                        try {
                            MongoClient = require('mongodb').MongoClient;
                        }
                        catch (e) {
                            throw new Error("You need to install \"mongodb\" package. It's required for checking ReplicaSet state.");
                        }
                        return [4 /*yield*/, MongoClient.connect(uris[0], {
                                useNewUrlParser: true,
                                useUnifiedTopology: true,
                            })];
                    case 2:
                        conn = _a.sent();
                        _a.label = 3;
                    case 3:
                        _a.trys.push([3, , 14, 16]);
                        return [4 /*yield*/, conn.db(this.opts.replSet.dbName)];
                    case 4:
                        db = _a.sent();
                        // MongoClient HACK which helps to avoid the following error:
                        //   "RangeError: Maximum call stack size exceeded"
                        //   in ./node_modules/mongodb/lib/core/sdam/topology_description.js:263:51
                        // Under the hood initiated server topology has type `RSGhost`
                        //   and this server does not support session
                        //   so for skipping the following code
                        //   https://github.com/mongodb/node-mongodb-native/blob/dc70c2de7d3dae2617708c45a1ea695d131e15f3/lib/operations/execute_operation.js#L33-L39
                        //   we just override shouldCheckForSessionSupport() method for current connection instance
                        db.topology.shouldCheckForSessionSupport = function () { return false; };
                        admin = db.admin();
                        members = uris.map(function (uri, idx) { return ({ _id: idx, host: db_util_1.getHost(uri) }); });
                        rsConfig = {
                            _id: this.opts.replSet.name,
                            members: members,
                            settings: __assign({ electionTimeoutMillis: 500 }, (this.opts.replSet.configSettings || {})),
                        };
                        _a.label = 5;
                    case 5:
                        _a.trys.push([5, 7, , 12]);
                        return [4 /*yield*/, admin.command({ replSetInitiate: rsConfig })];
                    case 6:
                        _a.sent();
                        return [3 /*break*/, 12];
                    case 7:
                        e_1 = _a.sent();
                        if (!(e_1 instanceof mongodb_1.MongoError && e_1.errmsg == 'already initialized')) return [3 /*break*/, 10];
                        log(e_1.errmsg + ": trying to set old config");
                        return [4 /*yield*/, admin.command({ replSetGetConfig: 1 })];
                    case 8:
                        oldConfig = (_a.sent()).config;
                        log('got old config:\n', oldConfig);
                        return [4 /*yield*/, admin.command({
                                replSetReconfig: oldConfig,
                                force: true,
                            })];
                    case 9:
                        _a.sent();
                        return [3 /*break*/, 11];
                    case 10: throw e_1;
                    case 11: return [3 /*break*/, 12];
                    case 12:
                        log('Waiting for replica set to have a PRIMARY member.');
                        return [4 /*yield*/, this._waitForPrimary()];
                    case 13:
                        _a.sent();
                        this.emit((this._state = 'running'));
                        log('running');
                        return [3 /*break*/, 16];
                    case 14: return [4 /*yield*/, conn.close()];
                    case 15:
                        _a.sent();
                        return [7 /*endfinally*/];
                    case 16: return [2 /*return*/];
                }
            });
        });
    };
    MongoMemoryReplSet.prototype._initServer = function (instanceOpts) {
        var serverOpts = {
            autoStart: false,
            binary: this.opts.binary,
            instance: instanceOpts,
            spawn: this.opts.replSet.spawn,
        };
        var server = new MongoMemoryServer_1.default(serverOpts);
        return server;
    };
    MongoMemoryReplSet.prototype._waitForPrimary = function (timeout) {
        if (timeout === void 0) { timeout = 30000; }
        return __awaiter(this, void 0, void 0, function () {
            var timeoutId, timeoutPromise;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        timeoutPromise = new Promise(function (resolve, reject) {
                            timeoutId = setTimeout(function () {
                                reject('Timed out in ' + timeout + 'ms. When waiting for primary.');
                            }, timeout);
                        });
                        return [4 /*yield*/, Promise.race(__spreadArrays(this.servers.map(function (server) {
                                var instanceInfo = server.getInstanceInfo();
                                if (!instanceInfo) {
                                    throw new Error('_waitForPrimary - instanceInfo not present ');
                                }
                                return instanceInfo.instance.waitPrimaryReady();
                            }), [
                                timeoutPromise,
                            ]))];
                    case 1:
                        _a.sent();
                        if (timeoutId != null) {
                            clearTimeout(timeoutId);
                        }
                        log('_waitForPrimary detected one primary instance ');
                        return [2 /*return*/];
                }
            });
        });
    };
    return MongoMemoryReplSet;
}(events_1.EventEmitter));
exports.default = MongoMemoryReplSet;
