// mongodb-memory-server-core/lib/util/__tests__/MongoBinary-test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var tmp = __importStar(require("tmp"));
var fs_1 = __importDefault(require("fs"));
var os_1 = __importDefault(require("os"));
var MongoBinary_1 = __importStar(require("../MongoBinary"));
var MongoBinaryDownload_1 = __importDefault(require("../MongoBinaryDownload"));
tmp.setGracefulCleanup();
jasmine.DEFAULT_TIMEOUT_INTERVAL = 600000;
var mockGetMongodPath = jest.fn().mockResolvedValue('/temp/path');
jest.mock('../MongoBinaryDownload', function () {
    return jest.fn().mockImplementation(function () {
        return { getMongodPath: mockGetMongodPath };
    });
});
describe('MongoBinary', function () {
    var tmpDir;
    beforeEach(function () {
        tmpDir = tmp.dirSync({ prefix: 'mongo-mem-bin-', unsafeCleanup: true });
    });
    // cleanup
    afterEach(function () {
        tmpDir.removeCallback();
        MongoBinaryDownload_1.default.mockClear();
        mockGetMongodPath.mockClear();
        MongoBinary_1.default.cache = {};
    });
    describe('getPath', function () {
        it('should get system binary from the environment', function () { return __awaiter(void 0, void 0, void 0, function () {
            var accessSpy;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        accessSpy = jest.spyOn(fs_1.default, 'access');
                        process.env.MONGOMS_SYSTEM_BINARY = '/usr/local/bin/mongod';
                        return [4 /*yield*/, MongoBinary_1.default.getPath()];
                    case 1:
                        _a.sent();
                        expect(accessSpy).toHaveBeenCalledWith('/usr/local/bin/mongod', expect.any(Function));
                        accessSpy.mockClear();
                        delete process.env.MONGOMS_SYSTEM_BINARY;
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe('getDownloadPath', function () {
        it('should download binary and keep it in cache', function () { return __awaiter(void 0, void 0, void 0, function () {
            var version, binPath;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        version = MongoBinary_1.LATEST_VERSION;
                        return [4 /*yield*/, MongoBinary_1.default.getPath({
                                downloadDir: tmpDir.name,
                                version: version,
                            })];
                    case 1:
                        binPath = _a.sent();
                        // eg. /tmp/mongo-mem-bin-33990ScJTSRNSsFYf/mongodb-download/a811facba94753a2eba574f446561b7e/mongodb-macOS-x86_64-3.5.5-13-g00ee4f5/
                        expect(MongoBinaryDownload_1.default).toHaveBeenCalledWith({
                            downloadDir: tmpDir.name,
                            platform: os_1.default.platform(),
                            arch: os_1.default.arch(),
                            version: version,
                        });
                        expect(mockGetMongodPath).toHaveBeenCalledTimes(1);
                        expect(MongoBinary_1.default.cache[version]).toBeDefined();
                        expect(MongoBinary_1.default.cache[version]).toEqual(binPath);
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe('getCachePath', function () {
        it('should get the cache', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        MongoBinary_1.default.cache['3.4.2'] = '/bin/mongod';
                        return [4 /*yield*/, expect(MongoBinary_1.default.getCachePath('3.4.2')).resolves.toEqual('/bin/mongod')];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe('getSystemPath', function () {
        it('should use system binary if option is passed.', function () { return __awaiter(void 0, void 0, void 0, function () {
            var accessSpy;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        accessSpy = jest.spyOn(fs_1.default, 'access');
                        return [4 /*yield*/, MongoBinary_1.default.getSystemPath('/usr/bin/mongod')];
                    case 1:
                        _a.sent();
                        expect(accessSpy).toHaveBeenCalledWith('/usr/bin/mongod', expect.any(Function));
                        accessSpy.mockClear();
                        return [2 /*return*/];
                }
            });
        }); });
    });
});
