// mongodb-memory-server-core/lib/util/__tests__/resolve-config-test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = __importDefault(require("fs"));
var tmp = __importStar(require("tmp"));
var util_1 = require("util");
var resolve_config_1 = __importStar(require("../resolve-config"));
tmp.setGracefulCleanup();
var mkdirAsync = util_1.promisify(fs_1.default.mkdir);
var writeFileAsync = util_1.promisify(fs_1.default.writeFile);
var outerPackageJson = {
    config: {
        mongodbMemoryServer: {
            version: '3.0.0',
        },
    },
};
var innerPackageJson = {
    config: {
        mongodbMemoryServer: {
            version: '4.0.0',
        },
    },
};
describe('resolveConfig', function () {
    var originalDir = process.cwd();
    var tmpObj;
    describe('configuration from closest package.json', function () {
        beforeAll(function () { return __awaiter(void 0, void 0, void 0, function () {
            var tmpName;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // Set up test project/subproject structure in a temporary directory:
                        //
                        //     project/
                        //     |-- subproject/
                        //     |   +-- package.json
                        //     +-- package.json
                        tmpObj = tmp.dirSync({ unsafeCleanup: true });
                        tmpName = tmpObj.name;
                        return [4 /*yield*/, mkdirAsync(tmpName + "/project")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, mkdirAsync(tmpName + "/project/subproject")];
                    case 2:
                        _a.sent();
                        // prettier-ignore
                        return [4 /*yield*/, Promise.all([
                                writeFileAsync(tmpName + "/project/package.json", JSON.stringify(outerPackageJson)),
                                writeFileAsync(tmpName + "/project/subproject/package.json", JSON.stringify(innerPackageJson)),
                            ])];
                    case 3:
                        // prettier-ignore
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        afterAll(function () {
            process.chdir(originalDir);
            tmpObj.removeCallback();
        });
        test('in project', function () {
            process.chdir(tmpObj.name + "/project");
            resolve_config_1.reInitializePackageJson();
            var got = resolve_config_1.default('VERSION');
            expect(got).toBe('3.0.0');
        });
        test('in subproject', function () {
            process.chdir(tmpObj.name + "/project/subproject");
            resolve_config_1.reInitializePackageJson();
            var got = resolve_config_1.default('VERSION');
            expect(got).toBe('4.0.0');
        });
        test('with explicit directory in reInitializePackageJson', function () {
            process.chdir(tmpObj.name + "/project");
            resolve_config_1.reInitializePackageJson(tmpObj.name + "/project/subproject");
            var got = resolve_config_1.default('VERSION');
            expect(got).toBe('4.0.0');
        });
    });
});
